Feature user register

Vasilijez opened the pull request:
I have introduced the following project folder structure. It is a little overcomplex, but I couldn’t overthink it anymore. If you don’t like this organization, I am ready to make changes. I remain flexible to make any necessary adjustments.

Files like `handler.go`, `route.go`, `service.go`, `repo.go` can become folders and be split into multiple files such as `auth_service.go` and `user_service.go`. But they can also stay as they are, thanks to the simplicity of our microservice.

The `middleware` folder can be populated with components such as logging, metrics, and others.

The `config` folder can contain all configuration files, like database settings and others.

I started writing tests in the test folder based on feature names, as this seemed like the best approach from my point of view.

___

I don’t have much to say about implementing user registration. It is CRUD.

I attempted to implement centralized error handling using `error_middleware.go`. It will intercept all HTTP requests and add HTTP status codes based on the errors. It also centralizes log data. For customized errors, it is handy to sometimes wrap them with additional data using `fmt.Errorf`.

magley committed:
feat: Add stub server

magley committed:
feat: Add dummy test function

magley committed:
feat: Add shell script for testing

magley committed:
feat: Containerize dummy app

magley committed:
feat: Create basic build workflow

magley committed:
feat: Add dependency caching to workflow

magley committed:
feat: Service connects to a database

magley committed:
docs: Update README.md

Vasilijez committed:
feat: Add `adminer`

Vasilijez committed:
feat: Add initial `User` model

Vasilijez committed:
feat: Add `container_name` for `adminer`

Vasilijez committed:
feat: Add `Address` model

Vasilijez committed:
feat: Improve `User` model

Vasilijez committed:
feat: Introduce `user_controller` and `user_routes`

Exported variables should be named using PascalCase.

Since I used `GORM` for modeling, I need a database connection. It uses `sql.DB` under the hood, so when I close `sql.DB`, I automatically close the `GORM` connection as well.

Vasilijez committed:
feat: Add `RegisterUser` controller

Vasilijez committed:
feat: Changing folder structure

This is quite a large commit—sorry about that! That said, most of the changes are related to restructuring the folder layout, which I believe is a crucial improvement.

I’ve decided to stop overthinking the project structure and went with something scalable, even if it's a bit overkill for now. I’ll explain more in the PR description.

Also, I’ve introduced error handling via middleware.

Vasilijez committed:
feat: Improve validation by leveraging `gin`

- I also tried to improve error handling.
- Typo: `ErrUserExists` → `ErrUsernameExists`

Vasilijez committed:
feat: Cover all packages under the `src` directory

Vasilijez committed:
test: Write unit tests for `Register` function

Vasilijez committed:
fix: Typo in `register_test`

Vasilijez committed:
Merge branch 'develop' into feature-user-register

magley commented:
Why are my commits part of this PR? Did you do this [e156f37](https://github.com/book-em/user-service/pull/7/commits/e156f3745f01756d0e2a5bce09534d2692463f9d) manually?

magley commented on src/main.go:
```diff
@@ -16,7 +23,15 @@ func Add(a int, b int) int {
+var (
+	server *gin.Engine
+	DB     *gorm.DB
+	RawDB  *sql.DB
```
If `RawDB` is used only in `main.go`, it should be private (lowercase).

Vasilijez replied:
https://github.com/book-em/user-service/pull/7/commits/186aa02e3d36aff0508df245d4c0153995a8dae4

magley commented on src/domain/user.go:
```diff
@@ -0,0 +1,21 @@
+	Email     string   `json:"email"    gorm:"type:varchar(60);not null;uniqueIndex"`
+	Name      string   `json:"name"     gorm:"type:varchar(60);not null;"`
+	Surname   string   `json:"surname"  gorm:"type:varchar(60);not null;"`
+	Role      UserRole `json:"role"     gorm:"type:varchar(5);not null;check:role IN ('guest','host','admin')"`
```
I'd prefer if user role was an integer enumeration. Besides, imposing `varchar(5)` seems very hacky.

Vasilijez replied:
We have spoken privately, so I’ll repeat what I told you, maybe it will be useful if our team grows in the future.

To be honest, what bothers me most about enumerations is the use of numbers, and then when I want to manually insert data into the database, I have to remember three numbers that represent the roles, and it's easy to make a mistake. On top of that, you have to define the type and the conversion of the enum values into strings.

As for using string values, with the `CHECK` constraint, I directly defined a restriction at the database schema level that prevents inserting anything other than allowed values. I used `VARCHAR(5)` just as an example of good practice, to set a max of 5 characters, rather than leaving it as just a variable-length string. Though, honestly, that’s not such a big deal.

It’s enough to define validation on the DTO level: `required, oneof=guest host admin`, and that wraps up the whole story about roles.

magley replied:
Fine by me, though I would never manually insert data into the database. It's almost always easier to write a helper method or to invoke the HTTP api directly (example: in MockerHub, I made that /dummy endpoint to fill up the DB with a bunch of users and empty repos. I used to do this even on Spring Boot, I never used `data.sql`).

magley commented on src/domain/address.go:
```diff
@@ -0,0 +1,10 @@
+package domain
+
+type Address struct {
+	ID      uint   `                gorm:"primaryKey"`
+	Street  string `json:"street"   gorm:"type:varchar(80);not null" binding:"required"`
+	City    string `json:"city"     gorm:"type:varchar(60);not null" binding:"required"`
+	State   string `json:"state"    gorm:"type:varchar(60)"`
+	Zip     string `json:"zip"      gorm:"type:varchar(20)"`
+	Country string `json:"country"  gorm:"type:varchar(80);not null" binding:"required"`
+}
```
This is overkill. `Address` should be a single string in the `User` table.

magley replied:
I say this because we almost certainly won't be doing any validation on each field. We do have a requirement to search by address... but that's for rooms only. And even then, we can just do a SQL Like operation.

Vasilijez replied:
Actually, I agree. https://github.com/book-em/user-service/pull/7/commits/9228bbb367e70774958b03c86b2850d38d6fb3c4

magley commented on src/api/route.go:
```diff
@@ -0,0 +1,24 @@
+	router := rg.Group("users")
+
+	router.Use(middleware.ErrorHandlingMiddleware())
+	router.POST("/register", ur.userHandler.RegisterUser)
```
Do we need the "users" group? This microservice is for users, so that's already implied.

Vasilijez replied:
No, we don't. This was an overkill. https://github.com/book-em/user-service/pull/7/commits/059a8c2b014f5a6f0c0ebb08453687ea9a3feda7 and https://github.com/book-em/user-service/pull/7/commits/36777586bd57265427a4f1dae08c6f6ea9bd1bae

magley commented on src/api/middleware/error_middleware.go:
```diff
@@ -0,0 +1,44 @@
+	default:
+		return http.StatusInternalServerError
+	}
+}
```
If this function is only used in this file, make it private.

Vasilijez replied:
https://github.com/book-em/user-service/pull/7/commits/186aa02e3d36aff0508df245d4c0153995a8dae4

Vasilijez commented:
Honestly, I don't know, maybe I accidentally created the `feature-user-register` branch based on PR #2 instead of `develop`, as I usually do.

magley reviewed (changes requested):
As far as the project structure goes, I think it's overengineered. There's no need to package by layer for such a small service. Also, do we really need a struct for each layer? I would've written everything as a function. So, registering users would be `HandleRegister(), Register(), DBRegister()`. I think that's closer to idiomatic Go. We don't want another MockerHub that's trying to be like Spring Boot.

It doesn't matter. Don't change it now, we would just be wasting time. I want to finish everything ASAP.

I didn't test it manually yet because we have no frontend nor a simple python script (like a `import requests; requests.post("...", {...})`, and the Address model is needlessly complex.

Vasilijez committed:
fix: Remove `address` model, use `string` instead

Vasilijez committed:
fix: Sometimes `backend` starts before `database`

We could add a backoff, but I think it's smarter to keep retrying indefinitely without delay, since we want all services to start as soon as possible.

Vasilijez committed:
fix: Remove `users` as group name in path

Vasilijez committed:
Refactor: Removed `User` prefix as only one service/repo is needed

I made this change based on our specification, as we likely won’t need more than one service or repository.

Vasilijez committed:
fix: Make lowercase all not exported variables and functions

I was unsure about the `Register` handler and decided to make it lowercase, as I currently don’t see a reason to export it. Integration tests will likely be performed via API calls.

Vasilijez commented:
I applied the following method for resolving the `rebase` problem. 

I created the `feature-user-register-rebase` branch and then used the `git cherry-pick` operation to copy my commits from the `feature-user-register` branch. This way, I avoided having @magley’s overlapping commits from the `develop` branch show up in the PR.

All previous commits remained unchanged and have the same hash values. All the commits I copied have new hash values. The `feature-user-register` branch will not be merged.

If I went with the `rebase` operation, it’s very likely that even @magley’s commits would get new hash values, and who knows what else would happen, since it would be necessary to go back through history and resolve conflicts, which already appear from the very first commit. That’s a very messy solution, and that’s why I’m trying to avoid handling it that way.

The simplest methods, like changing the base branch from `develop` to `master` and back, didn’t work, this was the next simplest approach.
